src/pcap/api.go

```diff
@@ -0,0 +1,650 @@
+			cancel(errorf(codes.InvalidArgument, "read operation: operation was nil: %w", errNilField))
+			return
+		}
+
+		// Gets a Stop message if it's there. Returns nil in any other case, also other message types.
+		stop := msg.GetStop()
+		if stop == nil {
+			cancel(errorf(codes.InvalidArgument, "read operation: expected message of type Stop: %w", errInvalidPayload))
+			return
+		}
+
+		// cancel without cause - normal exit
+		zap.L().Debug("client requested stop of capture")
+		cancel(nil)
+	}()
+}
+
+type streamPreparer interface {
+	prepareStreamToTarget(context.Context, *CaptureOptions, AgentEndpoint, credentials.TransportCredentials, *zap.Logger) (captureReceiver, error)
+}
+
+func (api *API) capture(ctx context.Context, stream responseSender, streamPrep streamPreparer, opts *CaptureOptions, targets []AgentEndpoint, creds credentials.TransportCredentials, log *zap.Logger) (<-chan *CaptureResponse, error) {
+	var captureCs []<-chan *CaptureResponse
+
+	runningCaptures := 0
+	for _, target := range targets {
+		log = log.With(zap.String("target", target.String()))
+		log.Info("starting capture")
+
+		captureStream, err := streamPrep.prepareStreamToTarget(ctx, opts, target, creds, log)
+		if err != nil {
+			errMsg := convertStatusCodeToMsg(err, target)
+			sendErr := stream.Send(errMsg)
+			if sendErr != nil {
+				// FIXME the return interrupts the for-loop over targets and returns in case of send error
+				return nil, sendErr
+			}
+
+			log.Info("capture cannot be started")
+
+			continue
+		}
+
+		runningCaptures++
+
+		log.Info("Add capture waiting group")
+
+		err = api.registerAgentCaptureStream(ctx, &captureStream, log)
+		if err != nil {
+			//TODO: implement error handling
+		}
+
+		c := readMsgFromStream(ctx, captureStream, target, api.bufConf.Size)
+		captureCs = append(captureCs, c)
+	}
+
+	if runningCaptures == 0 {
+		log.Error("Starting of all captures failed during stream preparation")
+		return nil, errorf(codes.FailedPrecondition, "Starting of all captures failed")
+	}
+
+	// merge channels to one channel and send to forward to stream
+	out := mergeResponseChannels(captureCs, api.bufConf.Size)
+	return out, nil
+}
+
+// starts the API drain process, where all ongoing captures are sent a Stop signal so they can terminate gracefully.
+// Runs with a drain timeout and will return an error if the context times out.
+func (api *API) Drain() error {
+	zap.L().Debug("Starting drain with timeout", zap.Duration("timeout", api.drainTimeout))
+	drainTimeout, cancel := context.WithTimeout(context.Background(), api.drainTimeout)
+	defer cancel()
+
+	drained := api.drainStreams()
+	for {
+		select {
+		case <-drained:
+			zap.L().Info("Drain completed successfully.")
+			return nil
+		case <-drainTimeout.Done():
+			zap.L().Warn("Drain timeout reached, but not all clients completed.")
+			return drainTimeout.Err()
+		}
+	}
+}
+
+func (api *API) drainStreams() chan struct{} {
+	defer api.captureLock.Unlock()
+	api.captureLock.Lock()
```

Comment:
Defer the unlock only after acquiring the lock.

Reply:
Resolved as we reimplemented drain logic.